src/solvers/dg/interpolation.jl

```diff
@@ -1,88 +1,220 @@
-        end
-        data_out[v, i, j] = acc
+        data_out[v, i, j, k] += matrix[i, ii] * matrix[j, jj] * matrix[k, kk] * data_in[v, ii, jj, kk]
       end
     end
   end
 
   return data_out
 end
 
+"""
+    multiply_dimensionwise(matrix::AbstractMatrix, data_in::AbstractArray{<:Any, NDIMS+1})
+
+Multiply the array `data_in` by `matrix` in each coordinate direction, where `data_in`
+is assumed to have the first coordinate for the number of variables and the remaining coordinates
+are multiplied by `matrix`.
+"""
+function multiply_dimensionwise(matrix::AbstractMatrix, data_in::AbstractArray{<:Any, 2})
+  # 1D
+  # optimized version of multiply_dimensionwise_naive
+  size_out = size(matrix, 1)
+  n_vars   = size(data_in, 1)
+  data_out = zeros(promote_type(eltype(data_in), eltype(matrix)), n_vars, size_out)
 
-# Interpolate data using the given Vandermonde matrix and return interpolated values (3D version).
-function interpolate_nodes(data_in::AbstractArray{T, 4},
-                           vandermonde, n_vars) where T
-  n_nodes_out = size(vandermonde, 1)
-  data_out = zeros(eltype(data_in), n_vars, n_nodes_out, n_nodes_out, n_nodes_out)
-  interpolate_nodes!(data_out, data_in, vandermonde, n_vars)
+  multiply_dimensionwise!(data_out, matrix, data_in)
+
+  return data_out
 end
 
-function interpolate_nodes!(data_out::AbstractArray{T, 4}, data_in::AbstractArray{T, 4},
-                            vandermonde, n_vars) where T
-  n_nodes_out = size(vandermonde, 1)
-  n_nodes_in  = size(vandermonde, 2)
+function multiply_dimensionwise(matrix::AbstractMatrix, data_in::AbstractArray{<:Any, 3})
+  # 2D
+  # optimized version of multiply_dimensionwise_naive
+  size_out = size(matrix, 1)
+  n_vars   = size(data_in, 1)
+  data_out = zeros(promote_type(eltype(data_in), eltype(matrix)), n_vars, size_out, size_out)
 
-  for k in 1:n_nodes_out, j in 1:n_nodes_out, i in 1:n_nodes_out
-    for v in 1:n_vars
-      acc = zero(eltype(data_out))
-      for kk in 1:n_nodes_in, jj in 1:n_nodes_in, ii in 1:n_nodes_in
-        acc += vandermonde[i, ii] * vandermonde[j, jj] * vandermonde[k, kk] * data_in[v, ii, jj, kk]
-      end
-      data_out[v, i, j, k] = acc
-    end
-  end
+  multiply_dimensionwise!(data_out, matrix, data_in)
 
   return data_out
 end
 
+function multiply_dimensionwise(matrix::AbstractMatrix, data_in::AbstractArray{<:Any, 4})
+  # 3D
+  # optimized version of multiply_dimensionwise_naive
+  size_out = size(matrix, 1)
+  n_vars   = size(data_in, 1)
+  data_out = zeros(promote_type(eltype(data_in), eltype(matrix)), n_vars, size_out, size_out, size_out)
+
+  multiply_dimensionwise!(data_out, matrix, data_in)
+
+  return data_out
+end
+
+
+# in-place version of multiply_dimensionwise
+function multiply_dimensionwise!(data_out::AbstractArray{<:Any, 2}, matrix::AbstractMatrix,
+                                 data_in ::AbstractArray{<:Any, 2})
+  @tullio threads=false data_out[v, i] = matrix[i, ii] * data_in[v, ii]
+
+  return nothing
+end
```

Comment:
Could you please also add a docstring here and indicate 1D/2D/3D in the comments? Also, it would be nice to have a comment on what the `@tullio` line does (and that it uses Einstein summation) such that one can understand this code without having to dig into `Tullio.jl`

Reply:
Sure.

Reply:
But I would hesitate to add a docstring here, since they are mainly made for internal use. If we add a doccstring, they become part of the public API. Do we really want that?

Reply:
No, you're right. A comment would suffice.